### PR TITLE
Quaternion smallest three compression

### PR DESCRIPTION
**Pull request by hamzzam**
## Summary

Quaternions are compressed using the smallest three compression before writing to network packets.

## References

https://gafferongames.com/post/snapshot_compression/

## Checklist
- [ ] CI/CD checks pass `npm run check`
  - [ ] Linter passing via `npm run lint`
  - [ ] Typescript passing via `npm run check-errors`
  - [ ] Unit & Integration tests passing via `npm run test`
  - [ ] Docker build process passing via `npm run build-client`
- [ ] If this PR is still a WIP, convert to a draft 
- [ ] When this PR is ready, mark it as "Ready for review"
- [ ] Changes have been manually QA'd
- [ ] Changes reviewed by at least 2 approved reviewers

## QA Steps

_List any additional steps required to QA the changes of this PR, as well as any supplemental images or videos._

## Reviewers
@HexaField @speigg

### CODE REVIEW

**Comment by speigg** [comment]
You can regain some precision by endowing the smallest three components in the range of 1/sqrt(2), i.e.,  [-0.707107,+0.707107] instead of [-1,+1]. See https://gafferongames.com/post/snapshot_compression/

File: packages/engine/src/networking/serialization/DataWriter.ts [hunk 1]
```diff
@@ -81,10 +86,99 @@ export const writeVector4 = (vector4: Vector4SoA) => (v: ViewCursor, entity: Ent
   return (changeMask > 0 && writeChangeMask(changeMask)) || rewind()
 }
 
+// Writes a compressed Quaternion value to the network stream. This function uses the "smallest three"
+// method, which is well summarized here: http://gafferongames.com/networked-physics/snapshot-compression/
+// Started with this code as baseline: https://gist.github.com/StagPoint/bb7edf61c2e97ce54e3e4561627f6582
+export const writeComporessedRotation = (vector4: Vector4SoA) => (v: ViewCursor, entity: Entity) => {
```
Comment (HexaField) [inline packages/engine/src/networking/serialization/DataWriter.ts:92]: typo, should be spelled `writeCompressedRotation`

Comment (hamzzam) [reply packages/engine/src/networking/serialization/DataWriter.ts]: Fixed.

File: packages/engine/src/networking/serialization/DataReader.ts [hunk 2]
```diff
@@ -3,6 +3,7 @@ import { TypedArray } from 'bitecs'
 import { NetworkId } from '@xrengine/common/src/interfaces/NetworkId'
 import { UserId } from '@xrengine/common/src/interfaces/UserId'
 
+import { FLOAT_PRECISION_MULT, QUAT_MAX_RANGE } from '../../common/constants/MathConstants'
```
Comment (speigg) [inline packages/engine/src/networking/serialization/DataReader.ts:6]: These constants aren't useful in any other context, so it seems more appropriate to define them somewhere in the scope of these serialization modules to minimize clutter.

File: packages/engine/src/networking/serialization/DataReader.ts [hunk 3]
```diff
@@ -78,10 +80,83 @@ export const readVector4 = (vector4: Vector4SoA) => (v: ViewCursor, entity: Enti
...
+// Reads a compressed rotation value from the network stream. This value must have been previously written
+// with WriteCompressedRotation() in order to be properly decompressed.
+export const readCompressedRotation = (vector4: Vector4SoA) => (v: ViewCursor, entity: Entity | undefined) => {
+  const changeMask = readUint8(v)
+  if (changeMask <= 0) return
+
+  let compressedBinaryData = readUint32(v)
+
+  const expand = (compressedBinaryData: number) => {
+    const valueReadMask = 0b00000000000000000000000111111111
```
Comment (speigg) [inline packages/engine/src/networking/serialization/DataReader.ts:92]: This function is being recreated every time readCompressedRotation is called, which creates unnecessary garbage and slows down performance.

File: packages/engine/src/networking/serialization/DataWriter.ts [hunk 4]
```diff
@@ -81,10 +85,115 @@ export const writeVector4 = (vector4: Vector4SoA) => (v: ViewCursor, entity: Ent
...
+
+    // Multiply with QUAT_MAX_RANGE & FLOAT_PRECISION_MULT before compression so that values are
+    // capped to required range([-0.707107,+0.707107]) and precision(three decimal places)
+    a *= sign * QUAT_MAX_RANGE * FLOAT_PRECISION_MULT
+    b *= sign * QUAT_MAX_RANGE * FLOAT_PRECISION_MULT
+    c *= sign * QUAT_MAX_RANGE * FLOAT_PRECISION_MULT
+
+    maxIndex = maxIndex | 0
+
+    const compress = (value: number) => {
```
Comment (speigg) [inline packages/engine/src/networking/serialization/DataWriter.ts:160]: This function is also be created anew every time writeCompressedRotation is called

File: packages/common/src/utils/mathUtils.ts [hunk 5]
```diff
@@ -77,3 +77,7 @@ export const isZero = (v: Vector3): boolean => {
 export const normalizeRange = (val, rangeMin, rangeMax) => {
   return (val - rangeMin) / (rangeMax - rangeMin)
 }
+
+export const approxeq = (v1, v2, epsilon = 0.001) => {
```
Comment (speigg) [inline packages/common/src/utils/mathUtils.ts:81]: can you rename to `equalsEpsillon` for clarity? Also, v1 and v2 parameters should be typed.

**Review by speigg — CHANGES_REQUESTED** [review]
Please write tests for readCompressedRotation and writeCompressedRotation

**Comment by hamzzam** [comment]
> Please write tests for readCompressedRotation and writeCompressedRotation

Done.

**Review by HexaField — APPROVED** [review]

**Pull request merged**